network/lwip2transport/udp.go

```diff
@@ -15,103 +15,110 @@
 package lwip2transport
 
 import (
-	"context"
 	"fmt"
 	"net"
 	"sync"
-	"time"
 
-	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
+	"github.com/Jigsaw-Code/outline-internal-sdk/network"
 	lwip "github.com/eycorsican/go-tun2socks/core"
 )
 
 // Compilation guard against interface implementation
 var _ lwip.UDPConnHandler = (*udpHandler)(nil)
+var _ network.PacketResponseReceiver = (*udpConnResponseWriter)(nil)
 
 type udpHandler struct {
-	// Protects the connections map
-	sync.Mutex
-
-	// Used to establish connections to the proxy
-	listener transport.PacketListener
-
-	// How long to wait for a packet from the proxy. Longer than this and the connection
-	// is closed.
-	timeout time.Duration
-
-	// Maps local UDP addresses (IPv4:port/[IPv6]:port) to connections to the proxy.
-	conns map[string]net.PacketConn
+	mu      sync.Mutex                             // Protects the sessions field
+	handler network.PacketProxy                    // A network stack neutral implementation of UDP handler
+	senders map[string]network.PacketRequestSender // Maps local lwIP UDP socket (IPv4:port/[IPv6]:port) to PacketSession
 }
 
 // newUDPHandler returns a lwIP UDP connection handler.
 //
-// `pl` is an UDP packet listener.
+// `h` is a handler that handles UDP packets.
 // `timeout` is the UDP read and write timeout.
-func newUDPHandler(pl transport.PacketListener, timeout time.Duration) *udpHandler {
+func newUDPHandler(h network.PacketProxy) *udpHandler {
 	return &udpHandler{
-		listener: pl,
-		timeout:  timeout,
-		conns:    make(map[string]net.PacketConn, 8),
+		handler: h,
+		senders: make(map[string]network.PacketRequestSender, 8),
 	}
 }
 
-func (h *udpHandler) Connect(tunConn lwip.UDPConn, target *net.UDPAddr) error {
-	proxyConn, err := h.listener.ListenPacket(context.Background())
+// Connect creats a new UDP session from the [transport.PacketServer]. It also schedule a timer with the timeout of the
+// session; the session will be closed once the timer triggers.
+func (h *udpHandler) Connect(tunConn lwip.UDPConn, _ *net.UDPAddr) error {
+	return h.newSession(tunConn)
+}
+
+// ReceiveTo relays packets from the lwIP TUN device to the proxy. It's called by lwIP.
+func (h *udpHandler) ReceiveTo(tunConn lwip.UDPConn, data []byte, destAddr *net.UDPAddr) error {
+	laddr := tunConn.LocalAddr().String()
+
+	h.mu.Lock()
+	reqSender, ok := h.senders[laddr]
+	h.mu.Unlock()
+
+	if !ok {
+		return fmt.Errorf("no session found for local address %v", laddr)
+	}
+	_, err := reqSender.WriteTo(data, destAddr)
```

Comment:
I'm still trying to figure out all the mux situation, and closing sequence. Perhaps we need a diagram of sorts. But I'll send the review for now.

Reply:
I feel it's OK, we should allow `Close` and `WriteTo` to be called simultaneously. Because for example, from `net.PacketConn`:

> Multiple goroutines may invoke methods on a PacketConn simultaneously.

The same requirement should be applied to `PacketRequestSender` as well. So the implementor needs to take care of this race condition. And since our `PacketListenerProxy` is already using a `PacketConn`, so this is taken care of automatically.